src/vmm/src/lib.rs

```diff
@@ -0,0 +1,405 @@
+        let kvm_api_ver = kvm.get_api_version();
+        if kvm_api_ver != KVM_API_VERSION as i32 {
+            return Err(Error::KvmApiVersion(kvm_api_ver));
+        }
+
+        // Create fd for interacting with kvm-vm specific functions.
+        let vm_fd = kvm.create_vm().map_err(Error::KvmIoctl)?;
+
+        let vmm = VMM {
+            vm_fd,
+            kvm,
+            guest_memory: GuestMemoryMmap::default(),
+            device_mgr: DeviceManager::new().map_err(Error::Device)?,
+            vcpus: vec![],
+        };
+
+        vmm.check_kvm_capabilities()?;
+
+        Ok(vmm)
+    }
+
+    /// Configure guest memory.
+    ///
+    /// # Arguments
+    ///
+    /// * `guest_mem_cfg` - [`MemoryConfig`](struct.MemoryConfig.html) struct containing guest
+    ///                     memory configurations.
+    pub fn configure_guest_memory(&mut self, guest_mem_cfg: MemoryConfig) -> Result<()> {
+        let mem_size = ((guest_mem_cfg.mem_size_mib as u64) << 20) as usize;
+
+        // Create guest memory regions.
+        // On x86_64, they surround the MMIO gap (dedicated space for MMIO device slots) if the
+        // configured memory size exceeds the latter's starting address.
+        let mem_regions = match mem_size.checked_sub(MMIO_MEM_START as usize) {
+            // Guest memory fits before the MMIO gap.
+            None | Some(0) => vec![(GuestAddress(0), mem_size)],
+            // Guest memory extends beyond the MMIO gap.
+            Some(remaining) => vec![
+                (GuestAddress(0), MMIO_MEM_START as usize),
+                (GuestAddress(FIRST_ADDR_PAST_32BITS), remaining),
+            ],
+        };
+
+        // Create guest memory from regions.
+        let guest_memory = GuestMemoryMmap::from_ranges(&mem_regions)
+            .map_err(|e| Error::Memory(MemoryError::VmMemory(e)))?;
+
+        if guest_memory.num_regions() > self.kvm.get_nr_memslots() {
+            return Err(Error::Memory(MemoryError::NotEnoughMemorySlots));
+        }
+
+        // Register guest memory regions with KVM.
+        guest_memory
+            .with_regions(|index, region| {
+                let memory_region = kvm_userspace_memory_region {
+                    slot: index as u32,
+                    guest_phys_addr: region.start_addr().raw_value(),
+                    memory_size: region.len() as u64,
+                    // It's safe to unwrap because the guest address is valid.
+                    userspace_addr: guest_memory.get_host_address(region.start_addr()).unwrap()
+                        as u64,
+                    flags: 0,
+                };
+
+                // Safe because the fd is a valid KVM file descriptor.
+                unsafe { self.vm_fd.set_user_memory_region(memory_region) }
+            })
+            .map_err(Error::KvmIoctl)?;
+
+        self.vm_fd
+            .set_tss_address(KVM_TSS_ADDRESS as usize)
+            .map_err(Error::KvmIoctl)?;
```

Comment:
From my understanding it's the other way around, but maybe I got it wrong - could you explain a bit why it's unnecessary?
Here's where I'm coming from - I know the guest boots without knowing the TSS addr but [this snippet](https://elixir.bootlin.com/linux/latest/source/arch/x86/kvm/vmx/vmx.c#L2858) when the guest enters real mode looks for it:
```c
static void enter_rmode(struct kvm_vcpu *vcpu)
{
     ...
	/*
	 * Very old userspace does not call KVM_SET_TSS_ADDR before entering
	 * vcpu. Warn the user that an update is overdue.
	 */
	if (!kvm_vmx->tss_addr)
		printk_once(KERN_WARNING "kvm: KVM_SET_TSS_ADDR need to be "
			     "called before entering vcpu\n");

    ...
	vmcs_write32(GUEST_TR_LIMIT, RMODE_TSS_SIZE - 1);
```

Reply:
If you look at the comment above, only "very old userspace does not call KVM_SET_TSS_ADDR" which means that for old hardware you need to manually call it. Also, looking at the `KVM_SET_TSS_ADDR` it seems that for unrestricted guests, this is a no-op: https://elixir.bootlin.com/linux/latest/source/arch/x86/kvm/vmx/vmx.c#L4630.

More about unrestricted guests in the Intel SDM chapter "25.6 UNRESTRICTED GUESTS": https://www.intel.com/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-vol-3c-part-3-manual.pdf

In a nutshell, we are running an "unrestricted guest" because we are setting both CR0.PE and CR0.PG to 1:
- [CR0.PE](https://github.com/aghecenco/vmm-reference/blob/67557522737f36b8e41f33194ee0945c3b810833/src/vmm/src/vcpu/mod.rs#L147)
- [CR0.PG](https://github.com/aghecenco/vmm-reference/blob/67557522737f36b8e41f33194ee0945c3b810833/src/vmm/src/vcpu/mod.rs#L176)

Because we are running unrestricted guests, there is no need to call SET_TSS. Does that make sense?

Reply:
Makes total sense. Thank you for the detailed explanation :blue_book:

Reply:
So obviously something was wrong in my explanation:

> In a nutshell, we are running an "unrestricted guest" because we are setting both CR0.PE and CR0.PG to 1:

KVM cares not of what we set in `CR0.PE` and `CR0.PG`. Instead it decides if `enable_unrestricted_guest` is true by (eventually, in a really long chain of functions) calling `rdmsr` (which is short for read msr) on the host and checking if `SECONDARY_EXEC_UNRESTRICTED_GUEST` is 1.

That being now corrected, it still seems like unrestricted guests is a feature that was introduced first in Westemere (according to my friend [wikipedia](https://en.wikipedia.org/wiki/X86_virtualization#Intel_virtualization_(VT-x))), which was released in 2010, so I still think we should be reasonably safe in deleting `SET_TSS_ADDR` because we're not planning on running on older hardware.